File: Day44/Part1/chrono_path.py
import collections

def solve():
    # Hardcoded Input
    N, M = 4, 4
    E = 10
    
    # Cost grid
    costs = [
        [0, 2, 1, 1],
        [1, 5, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 0]
    ]
    # Note: Cost to enter (0,0) is usually irrelevant as we start there, 
    # but let's assume we start with E and only pay when moving INTO a cell.
    # Or pay for start? Usually "enter cell". We are already at start.
    # Let's assume cost is paid when entering.
    
    # Time Gates: (r, c) -> set of forbidden times
    time_gates = {
        (1, 1): {2}, # Cannot enter (1,1) at t=2
        (0, 2): {2}  # Cannot enter (0,2) at t=2
    }
    
    start_pos = (0, 0)
    end_pos = (N-1, M-1)
    
    # BFS State: (r, c, energy, time)
    # Pruning: max_energy[r][c]
    
    max_energy = [[-1] * M for _ in range(N)]
    
    queue = collections.deque()
    
    # Initial state
    # Time 0. Energy E.
    # We are at (0,0).
    queue.append((0, 0, E, 0))
    max_energy[0][0] = E
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while queue:
        r, c, cur_e, cur_t = queue.popleft()
        
        if (r, c) == end_pos:
            print(f"Minimum time: {cur_t}")
            return
        
        next_t = cur_t + 1
        
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            
            if 0 <= nr < N and 0 <= nc < M:
                # Check Time Gate
                if (nr, nc) in time_gates and next_t in time_gates[(nr, nc)]:
                    continue
                
                # Check Energy
                move_cost = costs[nr][nc]
                next_e = cur_e - move_cost
                
                if next_e >= 0:
                    # Pruning
                    if next_e > max_energy[nr][nc]:
                        max_energy[nr][nc] = next_e
                        queue.append((nr, nc, next_e, next_t))
                        
    print("Impossible")

if __name__ == "__main__":
    solve()

File: learnLeetCode/Day44/Part1/chrono_path.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <map>

using namespace std;

struct State {
    int r, c;
    int energy;
    int time;
};

int main() {
    int N = 4, M = 4;
    int E = 10;

    vector<vector<int>> costs = {
        {0, 2, 1, 1},
        {1, 5, 1, 1},
        {1, 1, 1, 1},
        {1, 1, 1, 0}
    };

    // Map (r, c) -> set of forbidden times
    map<pair<int, int>, set<int>> time_gates;
    time_gates[{1, 1}] = {2};
    time_gates[{0, 2}] = {2};

    // max_energy[r][c]
    vector<vector<int>> max_energy(N, vector<int>(M, -1));

    queue<State> q;
    q.push({0, 0, E, 0});
    max_energy[0][0] = E;

    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};

    while (!q.empty()) {
        State cur = q.front();
        q.pop();

        if (cur.r == N - 1 && cur.c == M - 1) {
            cout << "Minimum time: " << cur.time << endl;
            return 0;
        }

        int next_t = cur.time + 1;

        for (int i = 0; i < 4; ++i) {
            int nr = cur.r + dr[i];
            int nc = cur.c + dc[i];

            if (nr >= 0 && nr < N && nc >= 0 && nc < M) {
                // Check Time Gate
                if (time_gates.count({nr, nc}) && time_gates[{nr, nc}].count(next_t)) {
                    continue;
                }

                int move_cost = costs[nr][nc];
                int next_e = cur.energy - move_cost;

                if (next_e >= 0) {
                    if (next_e > max_energy[nr][nc]) {
                        max_energy[nr][nc] = next_e;
                        q.push({nr, nc, next_e, next_t});
                    }
                }
            }
        }
    }

    cout << "Impossible" << endl;

    return 0;
}
